### PR TITLE
Avoid removing and manipulating event target list in go-routine

### LINKED ISSUE #6483: Minio server crashed with "concurrent map read and map write" fatal error
## Expected Behavior
Minio not crashing

## Current Behavior
```
   2 sep 18 17:10:14 xxxxx systemd[1]: Stopping Minio...
   3 sep 18 17:10:14 xxxxx minio[845]: Exiting on signal terminated
   4 sep 18 17:10:14 xxxxx minio[845]: fatal error: concurrent map read and map write
   5 sep 18 17:10:14 xxxxx minio[845]: goroutine 529816617 [running]:
   6 sep 18 17:10:14 xxxxx minio[845]: runtime.throw(0x1486267, 0x21)
   7 sep 18 17:10:14 xxxxx minio[845]: /opt/go/src/runtime/panic.go:616 +0x81 fp=0xc42c046e88 sp=0xc42c046e68 pc=0x42ae41
   8 sep 18 17:10:14 xxxxx minio[845]: runtime.mapaccess2(0x1265860, 0xc420698c60, 0xc42c046f88, 0xc428a94480, 0xc423422a80)
   9 sep 18 17:10:14 xxxxx minio[845]: /opt/go/src/runtime/hashmap.go:409 +0x225 fp=0xc42c046ed0 sp=0xc42c046e88 pc=0x4077d5
  10 sep 18 17:10:14 xxxxx minio[845]: github.com/minio/minio/pkg/event.(*TargetList).Remove.func1(0xc42c557e60, 0xc427717420, 0x1, 0x1, 0xc4205bb120)
  11 sep 18 17:10:14 xxxxx minio[845]: /q/.q/sources/gopath/src/github.com/minio/minio/pkg/event/targetlist.go:79 +0x134 fp=0xc42c046fb8 sp=0xc42c046ed0 pc=0xb205c4
  12 sep 18 17:10:14 xxxxx minio[845]: runtime.goexit()
  13 sep 18 17:10:14 xxxxx minio[845]: /opt/go/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc42c046fc0 sp=0xc42c046fb8 pc=0x458491
  14 sep 18 17:10:14 xxxxx minio[845]: created by github.com/minio/minio/pkg/event.(*TargetList).Remove
  15 sep 18 17:10:14 xxxxx minio[845]: /q/.q/sources/gopath/src/github.com/minio/minio/pkg/event/targetlist.go:74 +0xc3
```

## Possible Solution
Do not know, but it looks like `mc mirror -w` can be responsible for this

## Steps to Reproduce (for bugs)
Not sure yet, we need to ask @curantes if there is a reliable way to reproduce this.

## Context
Fixing bug

## Regression
Do not know yet

## Your Environment
minio RELEASE.2018-09-01T00-38-25Z

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Avoid removing and manipulating event target list in go-routine

## Motivation and Context
Fixes #6483

## Regression
No

## How Has This Been Tested?
- Couldn't reproduce original issue
- This fix is a theoretical possibility which exists in the code

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

File: pkg/event/targetlist.go [hunk 1]
```diff
@@ -74,24 +74,15 @@ func (list *TargetList) Remove(targetids ...TargetID) <-chan TargetIDErr {
 	go func() {
```
Comment (balamurugana) [inline pkg/event/targetlist.go:74]: Can't understand the fix.

Comment (harshavardhana) [reply pkg/event/targetlist.go:74]: It's a theoretical fix i.e avoid concurrent map read/write @balamurugana as indicated in the issue #6483

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#c0d83e47-448b-4f15-913c-d09a5723fa6a) |
<a name="c0d83e47-448b-4f15-913c-d09a5723fa6a"></a>
### 6485-8ae6d3d/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30042
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp af9552287fbf:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 2 minutes and 39 seconds
(2/13) Running aws-sdk-go tests ... done in 2 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 31 seconds
(5/13) Running aws-sdk-ruby tests ... done in 30 seconds
(6/13) Running mc tests ... done in 5 minutes and 15 seconds
(7/13) Running minio-dotnet tests ... FAILED in 5 minutes and 25 seconds
{
  "name": "minio-dotnet: CopyObject_Test5",
  "function": "Task<CopyObjectResult> CopyObjectAsync(string bucketName, string objectName, string destBucketName, string destObjectName = null, CopyConditions copyConditions = null, CancellationToken cancellationToken = default(CancellationToken))",
  "args": {
    "bucketName": "miniodotnet1dmvrquffxzrewr",
    "objectName": "miniodotnet42kkwlxpzp",
    "destBucketName": "miniodotnethuinuf8tw52rnof",
    "destObjectName": "miniodotnetyueqoq9lrg",
    "data": "6MB",
    "size": "6MB"
  },
  "duration": 5295,
  "status": "FAIL",
  "alert": "",
  "message": "Minio API responded with message=A header you provided implies functionality that is not implemented",
  "error": "A header you provided implies functionality that is not implemented: Minio.Exceptions.MinioException: Minio API responded with message=A header you provided implies functionality that is not implemented\n   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 362\n   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 504\n   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 350\n   at Minio.MinioClient.CopyObjectRequestAsync(String bucketName, String objectName, String destBucketName, String destObjectName, CopyConditions copyConditions, Dictionary`2 customHeaders, String resource, CancellationToken cancellationToken, Type type) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 986\n   at Minio.MinioClient.MultipartCopyUploadAsync(String bucketName, String objectName, String destBucketName, String destObjectName, CopyConditions copyConditions, Int64 copySize, Dictionary`2 metadata, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 1045\n   at Minio.MinioClient.CopyObjectAsync(String bucketName, String objectName, String destBucketName, String destObjectName, CopyConditions copyConditions, Dictionary`2 metadata, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 939\n   at Minio.Functional.Tests.FunctionalTest.CopyObject_Test5(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1059"
}

Executed 6 out of 13 tests successfully.
```

File: pkg/event/targetlist.go (hunk 1 above)
Comment (balamurugana) [reply pkg/event/targetlist.go:74]: where is concurrent read/write happening?

Comment (harshavardhana) [reply pkg/event/targetlist.go:74]:
```
   4 sep 18 17:10:14 xxxxx minio[845]: fatal error: concurrent map read and map write
   5 sep 18 17:10:14 xxxxx minio[845]: goroutine 529816617 [running]:
   6 sep 18 17:10:14 xxxxx minio[845]: runtime.throw(0x1486267, 0x21)
   7 sep 18 17:10:14 xxxxx minio[845]: /opt/go/src/runtime/panic.go:616 +0x81 fp=0xc42c046e88 sp=0xc42c046e68 pc=0x42ae41
   8 sep 18 17:10:14 xxxxx minio[845]: runtime.mapaccess2(0x1265860, 0xc420698c60, 0xc42c046f88, 0xc428a94480, 0xc423422a80)
   9 sep 18 17:10:14 xxxxx minio[845]: /opt/go/src/runtime/hashmap.go:409 +0x225 fp=0xc42c046ed0 sp=0xc42c046e88 pc=0x4077d5
  10 sep 18 17:10:14 xxxxx minio[845]: github.com/minio/minio/pkg/event.(*TargetList).Remove.func1(0xc42c557e60, 0xc427717420, 0x1, 0x1, 0xc4205bb120)
  11 sep 18 17:10:14 xxxxx minio[845]: /q/.q/sources/gopath/src/github.com/minio/minio/pkg/event/targetlist.go:79 +0x134 fp=0xc42c046fb8 sp=0xc42c046ed0 pc=0xb205c4
  12 sep 18 17:10:14 xxxxx minio[845]: runtime.goexit()
  13 sep 18 17:10:14 xxxxx minio[845]: /opt/go/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc42c046fc0 sp=0xc42c046fb8 pc=0x458491
  14 sep 18 17:10:14 xxxxx minio[845]: created by github.com/minio/minio/pkg/event.(*TargetList).Remove
  15 sep 18 17:10:14 xxxxx minio[845]: /q/.q/sources/gopath/src/github.com/minio/minio/pkg/event/targetlist.go:74 +0xc3
```

So basically line number 74 and 79

Comment (balamurugana) [reply pkg/event/targetlist.go:74]: The fix does not change anything between line number 74 and 79.  It looks like `TargetList.Remove()` is called in parallel somewhere. that needs to be fixed.

Comment (balamurugana) [reply pkg/event/targetlist.go:74]: Regression due to https://github.com/minio/minio/pull/5912

**Comment by nitisht** [comment]
ping @harshavardhana

**Comment by harshavardhana** [comment]
closing this PR

**Pull request closed without merging**